src/matching.js

```diff
@@ -0,0 +1,59 @@
+// TODO: we need some tests for this one
+export function pick(pathname, ...routes) {
+  for (const [path, onMatch] of routes) {
+    if (typeof path !== 'string') throw new Error(`Unexpected non-string value '${path}' in routes`)
+    const matched = match(pathname, path)
+    if (matched) return callOrReturn(onMatch, matched)
+  }
+  return null
+}
+
+export function callOrReturn(x, ...args) {
+  return typeof x === 'function' ? x(...args) : x
+}
+
+match.cache = {}
```

Comment:
Just curious, but why is this a property of the function below instead of a separate object? Is it to make a connection between the function and the object? 

It feels a bit convoluted: I need to think twice before I understand what's going on.

Reply:
Great question. Over the years it has gradually changed and apparently I ended up doing this. Originally it was like this:

```js
function test() {
  const cache = test.cache || (test.cache = {})
  if (test.cache && test.cache[key]) ...

  return test.cache[key] = ... 
}
```

Making the outside world oblivious to the fact that caching was used. This also makes it very portable.

The downside of this original approach is that typescript does not know about the `cache` property. The easiest route to keep the red squiggles at bay was to move `test.cache = {}` to it's separate line. The upside of this move is that it's instantly clear that this function has a form of caching, which has the following benefits:
- Can (almost) safely be called very often
- Returns stable values (handy when used in conjunction with hooks)

So all in all I think it is better to have `match.cache = {}` than `const matchCache = {}`

We could however think about introducing a utility (memoization), this however requires us to do this:

```js
export { 
  match: momoize(match)
}
```

And in some case (haven't checked this case) things become more problematic and less performant.